### DISCUSSION TITLE
Accessing from Java?

### DISCUSSION THREAD

**Discussion by jason53** [moov-io/ach#1373] [General]
I am interested in using this library but my application is written in Java.  Is there a Java version of this library?  Is JNI my only choice?  I realize this project offers a HTTP server but that option doesn't make sense with our architecture.  Thank you.

**Comment by adamdecaf** [comment]
Yea JNI is your only choice (that I know of) to call the native methods. A while ago I [was trying out Go's JNI support](https://github.com/moov-io/ach/pull/1327) and got it partially working. Do you know of a better option?

**Reply by jason53** [reply]
Thanks for the response.  I know that Java 21 has a replacement for JNI that is supposed to be better and more reliable.  I don't know of a better option.  From your response it sounds like Java is not on your roadmap :(

Does the HTTP server that is part of this project support secure endpoints?  In other words, requiring authenticated requests.  For example, requiring all the endpoints to use oAuth2 or Basic Auth.

**Reply by adamdecaf** [reply]
We would support Java interop if possible and I can help setup something for Java users. 

Currently the HTTP server does not require auth, but you can bind it to the local machine. Why do you need authenticated endpoints vs running it in a network confined setup?

**Reply by jason53** [reply]
Great to hear that you would support JNI access but to be clear I was asking if a pure Java library is on the roadmap.

Our information security department is typically very strict so I am anticipating them saying that network constraints are not enough and that the endpoints should require authentication.

**Reply by adamdecaf** [reply]
We can't support a Java port at this time. Moov is 100% Go so there's not a lot of space for other languages.

Do you know what auth methods they'd be looking for?

**Reply by jason53** [reply]
I understand the issue w/ supporting Java but thought I would ask.

For Auth method I would say a oAuth2 token based authentication ideally.  At a minimum it would support Basic auth.

**Reply by adamdecaf** [reply]
Have you found out more if this is required? It's more of a lift to add in the ACH server.

**Reply by jason53** [reply]
yes, it is required but we understand the effort involved to adding this functionality.  i think if we were to use moov-io/ach our integration preference would first be direct invocation w/ JNI and our second choice would be using the ACH server.  if we were to use the ACH server and it did not have any authentication built-in we would put something in front of it (either something we build or existing infrastructure appliance) that would require authentication.  hope that answers your question.

**Comment by adamdecaf** [comment]
We're working on a JNI interop package https://github.com/moov-io/ach/pull/1380

**Reply by jason53** [reply]
Great!  Thank you for your efforts!
